src/service.rs

```diff
@@ -67,57 +85,93 @@ impl<P: Platform> ServiceResources<P> {
+    }
+
+    pub fn platform_mut(&mut self) -> &mut P {
+        &mut self.platform
+    }
 }
 
-pub struct Service<P>
+pub struct Service<P, D = CoreOnly>
 where
     P: Platform,
+    D: DispatchRequirement<P>,
 {
-    eps: Vec<ServiceEndpoint, { MAX_SERVICE_CLIENTS::USIZE }>,
+    eps: Vec<ServiceEndpoint<D::BackendId, D::Context>, { MAX_SERVICE_CLIENTS::USIZE }>,
     resources: ServiceResources<P>,
+    dispatch: D,
 }
 
 // need to be able to send crypto service to an interrupt handler
-unsafe impl<P: Platform> Send for Service<P> {}
+unsafe impl<P: Platform, D: DispatchRequirement<P>> Send for Service<P, D> {}
 
 impl<P: Platform> ServiceResources<P> {
-    pub fn certstore(&mut self, client_ctx: &ClientContext) -> Result<ClientCertstore<P::S>> {
+    pub fn certstore(&mut self, ctx: &CoreContext) -> Result<ClientCertstore<P::S>> {
         self.rng()
-            .map(|rng| ClientCertstore::new(client_ctx.path.clone(), rng, self.platform.store()))
+            .map(|rng| ClientCertstore::new(ctx.path.clone(), rng, self.platform.store()))
             .map_err(|_| Error::EntropyMalfunction)
     }
 
-    pub fn counterstore(&mut self, client_ctx: &ClientContext) -> Result<ClientCounterstore<P::S>> {
+    pub fn counterstore(&mut self, ctx: &CoreContext) -> Result<ClientCounterstore<P::S>> {
         self.rng()
-            .map(|rng| ClientCounterstore::new(client_ctx.path.clone(), rng, self.platform.store()))
+            .map(|rng| ClientCounterstore::new(ctx.path.clone(), rng, self.platform.store()))
             .map_err(|_| Error::EntropyMalfunction)
     }
 
-    pub fn filestore(&mut self, client_ctx: &ClientContext) -> ClientFilestore<P::S> {
-        ClientFilestore::new(client_ctx.path.clone(), self.platform.store())
+    pub fn filestore(&mut self, ctx: &CoreContext) -> ClientFilestore<P::S> {
+        ClientFilestore::new(ctx.path.clone(), self.platform.store())
     }
 
     pub fn trussed_filestore(&mut self) -> ClientFilestore<P::S> {
         ClientFilestore::new(PathBuf::from("trussed"), self.platform.store())
     }
 
-    pub fn keystore(&mut self, client_ctx: &ClientContext) -> Result<ClientKeystore<P::S>> {
+    pub fn keystore(&mut self, ctx: &CoreContext) -> Result<ClientKeystore<P::S>> {
         self.rng()
-            .map(|rng| ClientKeystore::new(client_ctx.path.clone(), rng, self.platform.store()))
+            .map(|rng| ClientKeystore::new(ctx.path.clone(), rng, self.platform.store()))
             .map_err(|_| Error::EntropyMalfunction)
     }
 
+    pub fn dispatch<D: DispatchRequirement<P>>(
+        &mut self,
+        dispatch: &mut D,
+        backend: &BackendId<D::BackendId>,
+        ctx: &mut Context<D::Context>,
+        request: &Request,
+    ) -> Result<Reply, Error> {
+        #[cfg(feature = "ext")]
+        if let Request::Extension(request) = &request {
+            if let BackendId::Custom(backend) = backend {
+                return D::ExtensionId::try_from(request.id)
```

Comment:
Unless I'm missing something, I think this entire PR is "just" factoring this `D::ExtensionId::try_from(request.id)` call from the actual backend upstream to the service.